[TAN-5788] - Enable admins to set custom ARIA labels and descriptions for Screen-reader

amirabayoumi opened the pull request:
# Description
    - Allow admins to customize aria-label and aria-describedby for charts in the Reports Builder
    - Add accessibility description and label fields in report settings so charts on front pages include both attributes
    - Create new AccessibilityInputs component in chart widget settings
    - Add hook to pass accessibility inputs as props and link them with chart aria attributes  
    - Hide AccessibilityInputs when widget view is set to table

# Changelog
## Fixed
- [TAN-5788] - Enable admin defined aria-label and aria-describedby for charts

amirabayoumi committed:
WIP - Enable admins to set custom ARIA labels and descriptions for screen readers

amanda-anderson commented on front/app/containers/Admin/reporting/components/ReportBuilder/Widgets/ChartWidgets/VisitorsWidget/VisitorsCard/Narrow.tsx:
```diff
@@ -62,7 +79,16 @@ const Narrow = ({
+      {/* Chart container with ARIA attributes */}
+      <Box
+        role="img"
+        aria-label={localizedAriaLabel || 'Visitors chart'}
```
I know this is still WIP so you were probably planning to do it later, but just a small reminder to add this "Visitors chart" copy to a messages.ts file so it can be translated :relaxed:

amanda-anderson commented on front/app/containers/Admin/reporting/components/ReportBuilder/Widgets/ChartWidgets/VisitorsWidget/VisitorsCard/Narrow.tsx:
```diff
@@ -80,6 +106,13 @@ const Narrow = ({
         />
       </Box>
+
+      {/* description for screen readers */}
```
Since technically this is in the UI as well, we can probably remove this comment :) Or just say:

{/* description for accessibility */}

amanda-anderson commented on front/app/containers/Admin/reporting/components/ReportBuilder/Widgets/ChartWidgets/messages.ts:
```diff
@@ -67,4 +67,30 @@ export default defineMessages({
+  analyticsChartDescriptionTooltip: {
+    id: 'app.containers.admin.ReportBuilder.charts.analyticsChartDescriptionTooltip',
+    defaultMessage:
+      'Provide a more detailed description of the chart for screen reader users.',
```
Since this description will also be visible in the UI, could we update this to something like:

_"Provide a more detailed description of the chart and any relevant trends for improved accessibility"._

Also, when you update copy after it's been committed to your branch, you'll also have to change the key on line 88 so it registers as a "new" version:

E.g. Update it to:
id: 'app.containers.admin.ReportBuilder.charts.analyticsChartDescriptionTooltip2',

amirabayoumi replied:
Thanks @amanda-anderson  for giving me your feedback👌 appreciate it!

Commit:
Translations updated by CI (extract-intl)

amirabayoumi committed:
add ChartAccessibilityContext and useRechartsAccessibility hook

Commit:
Translations updated by CI (extract-intl)

amirabayoumi committed:
remove unused default exports

amirabayoumi committed:
show accessibility inputs for phase reports (not global reports)

amirabayoumi committed:
Hide AccessibilityInputs when widget view is set to table

amirabayoumi committed:
Merge branch 'master' into TAN-5788-Complex-graphics-for-Accessibility

amirabayoumi committed:
show accessibility inputs in all reports ( not only for phase)

IvaKop commented on front/app/components/admin/Graphs/LineChart/index.tsx:
```diff
@@ -17,6 +17,7 @@ import {
 
 import { isNilOrError } from 'utils/helperUtils';
 
+import { useRechartsAccessibility } from '../../../../containers/Admin/reporting/components/ReportBuilder/Widgets/ChartWidgets/_shared/useRechartsAccessibility';
```
An import like this is a clear sign your code is in the wrong place. Move this hook closer to where you use it, perhaps in the Graphs folder.

IvaKop commented on front/app/containers/Admin/reporting/components/ReportBuilder/Widgets/ChartWidgets/DemographicsWidget/Settings.tsx:
```diff
@@ -25,6 +25,7 @@ import {
 
 import platformTemplateMessages from '../../../Templates/PlatformTemplate/messages';
 import UserFieldSelect from '../../_shared/UserFieldSelect';
+import { AccessibilityInputs } from '../_shared/AccessibilityInputs';
```
By convention, we do default exports as opposed to named ones. This is not super important but it would be more consistent if you did the same.

IvaKop commented:
@amirabayoumi I see you didn't translate your branch. Every time there are new translation strings added, you need to add your branch to Crowdin so it can be translated. Click on [Check translation progress](https://epic.citizenlab.co/translations/TAN-5788-Complex-graphics-for-Accessibility) (also linked above) to create the branch. Then before merging to master, you need to check that all translations are done. Look into the deep dive video again, if needed, to remember the process.

IvaKop commented on front/app/components/admin/Graphs/LineChart/index.tsx:
```diff
@@ -57,6 +58,8 @@ const LineChart = <Row,>({
     LegendDimensions | undefined
   >();
 
+  const accessibilityProps = useRechartsAccessibility();
```
This hook requires `ChartAccessibilityProvider` to work. I am not convinced that is all instances this one and the rest of the components is used now and in the future this context will be available.

Can you provide some more information about why you chose to implement this via context. I don't think we normally do this for CraftJS components so more info will be appreciated.

IvaKop commented on front/app/containers/Admin/reporting/components/ReportBuilder/Widgets/ChartWidgets/_shared/useRechartsAccessibility.tsx:
```diff
@@ -0,0 +1,16 @@
+/**
+ * Hook to get accessibility props(role, aria-label and aria-describedby) for Recharts components
+ */
+export const useRechartsAccessibility = () => {
```
Why do you need such a complicated mechanism to access these props? Can you see how similar props are accessed in Chart components? Let's try and use a similar approach.

IvaKop reviewed (changes requested):
Unless there is a really good reason to use React Context for this particular use case, I believe this implementation might be simplified significantly by applying the same approach we normally do for accessing props in CraftJs components. Let's discuss in a pair coding session, if needed.

amirabayoumi committed:
remove useContext for chart accessibility

IvaKop commented on front/app/components/admin/Graphs/LineChart/index.tsx:
```diff
@@ -49,14 +50,24 @@ const LineChart = <Row,>({
     LegendDimensions | undefined
   >();
 
+  const accessibilityProps = {
```
Why is this structured differently than the other ones? Don't we also need role for the other graphs and accessibility layer in the other graphs?

amirabayoumi replied:
It is the same on all chart files ( if the chart component is the last child )
convert the props to attribute like that  
```
const accessibilityProps = {
    accessibilityLayer: true,
    role: 'img',
    'aria-label': ariaLabel,
    'aria-describedby': ariaDescribedBy,
    tabIndex: 0,
  };
```

amanda-anderson replied:
From what I understand, I also think Amira's changes are correct here :heavy_check_mark: In some of our other chart components, they aren't using the Recharts components directly, but more of our own components. 

We only need to use these "extended" accessibility props when a Recharts component is being used (I.e. We're at the "bottom" of the tree and are using the Recharts component directly).

amanda-anderson replied:
@amirabayoumi Since we do use this same codeblock often across components:
```
  const accessibilityProps = {
    accessibilityLayer: true,
    role: 'img',
    'aria-label': ariaLabel,
    'aria-describedby': ariaDescribedBy,
    tabIndex: 0,
  };
  ```
  
 I'm wondering if we can reduce code duplication though?
 
 Maybe in `Graphs/typings.ts` we can add something like:
 ```
 export interface RechartsAccessibilityProps {
  accessibilityLayer?: boolean;
  role?: string;
  'aria-label'?: string;
  'aria-describedby'?: string;
  tabIndex?: number;
}
 ```
 
 And then in the `Graphs/utils.ts` a little util function like:
 
 ```
 export const getRechartsAccessibilityProps = (
  ariaLabel: string,
  ariaDescribedBy: string
): RechartsAccessibilityProps => {
  return {
    accessibilityLayer: true,
    role: 'img',
    'aria-label': ariaLabel,
    'aria-describedby': ariaDescribedBy,
    tabIndex: 0,
  };
};
 ```
 
 And then we can use it in the components   `getRechartsAccessibilityProps(ariaLabel, ariaDescribedBy)`
 
 What do you think?? :thinking:

IvaKop commented on front/app/containers/Admin/reporting/components/ReportBuilder/Widgets/_shared/Card.tsx:
```diff
@@ -12,6 +12,11 @@ import useLocalize from 'hooks/useLocalize';
 
+import { useIntl } from 'utils/cl-intl';
+
+// import { ChartAccessibilityProvider } from './../ChartWidgets/_shared/ChartAccessibilityContext';
```
Missed comment here

IvaKop commented on front/app/containers/Admin/reporting/components/ReportBuilder/Widgets/_shared/Card.tsx:
```diff
@@ -37,8 +44,37 @@ const Container = ({ pagebreak, children, ...props }: ContainerProps) => {
+    ariaDescribedBy: chartAriaDescribedBy,
+  };
+
+  const childrenWithProps = React.Children.map(children, (child) => {
```
This needs to be deleted, it is not used, correct?

IvaKop replied:
Also all of the logic above should not be needed as we are passing the accessibility props directly to the components. Please revise this section once more to make sure you don't have code that is not necessary anymore.

amirabayoumi replied:
Done

IvaKop reviewed (changes requested):
It is looking so much better. Still some cleanup is necessary, though. Importantly, the logic for cloning the children should be removed as it should not be necessary with the new changes.

amirabayoumi committed:
delete commented import line

amirabayoumi committed:
remove cloneElement and add description text component

amirabayoumi committed:
fix descriptionId in vistor widget

amirabayoumi committed:
remove unused prop

amanda-anderson commented on front/app/containers/Admin/reporting/components/ReportBuilder/Widgets/ChartWidgets/MethodsUsedWidget/index.tsx:
```diff
@@ -1,23 +1,53 @@
 import React from 'react';
 
+import useLocalize from 'hooks/__mocks__/useLocalize';
```
Something to watch out for - this is actually the wrong import for `useLocalize`. the `hooks/__mocks__` are used only for testing purposes, and using this import here causes a crash in the app if you try and add a MethodsUsedWidget to a report.

It just needs to be updated to: ```import useLocalize from 'hooks/useLocalize';``` instead :relaxed: 

Sometimes the recommended import that VS code suggests is incorrect, so just good to double-check :)

amirabayoumi replied:
oops! yes i normally changed all other import but i missed this one

amanda-anderson commented on front/app/containers/Admin/reporting/components/ReportBuilder/Widgets/ChartWidgets/messages.ts:
```diff
@@ -67,4 +67,30 @@ export default defineMessages({
+  analyticsChartAriaLabelTooltip2: {
+    id: 'app.containers.admin.ReportBuilder.charts.analyticsChartAriaLabelTooltip2',
+    defaultMessage:
+      'Provide a label that describes the chart for improved accessibility',
```
Minor: For any sentences, our convention is to end with a full stop, e.g. 

"Provide a label that describes the chart for improved accessibility."
"Provide a more detailed description of the chart and any relevant trends for improved accessibility."

You might still see sentences without periods in our app, but those are usually old strings before we aligned on this convention, or strings that weren't caught before release :)

amanda-anderson commented on front/app/containers/Admin/reporting/components/ReportBuilder/Widgets/ChartWidgets/MethodsUsedWidget/index.tsx:
```diff
@@ -1,23 +1,53 @@
+      description={description}
+      pagebreak
+    >
+      <MethodsUsedCard {...props} {...accessibilityProps} />
```
Are the `accessibilityProps` being used in `MethodsUsedCard`? I'm seeing an error locally around types, but I don't see them being used in the chart itself.

I also don't see any Accessibility Settings for the `MethodsUsedWidget`, which makes sense since it's not a graph, so maybe we don't need these `accessibilityProps` at all in this `MethodsUsedWidget` component?

amirabayoumi replied:
Yes i just sent now in DM , MethodsUsedWidget contain a Statistic not a chart ! it was by mistake to add accessibilityProps here!!

amanda-anderson reviewed (changes requested):
Looking good so far!! :smile: Left a couple comments about the code, and then had 2 things from my functional review:

1. Can we add the Accessibility Settings to the "Survey Question" & "Community Monitor Health Score" widgets as well?

2. For the "Traffic Sources" widget, if we have it in Chart view, then add an aria description, then change it to Table view, the description still shows (when it shouldn't since we don't have the graph anymore)

amirabayoumi committed:
add accessibilityInput to HealthScoreChart widget

Commit:
Translations updated by CI (extract-intl)

amirabayoumi committed:
Add getRechartsAccessibilityProps helper function

amirabayoumi commented:
@amanda-anderson @IvaKop 

here is the last update => 

- created a shared getRechartsAccessibilityProps helper function so there are no repeated code in charts anymore! 
- added Accessibility setting to Community Monitor Health Score widget and for Question Survey will do it separately in another ticket and branch 
- removed the logic for cloning the children and using simple props passing to all children
- cleaned up all the files ( import mistakes and commented line ) 

hope to hear from you if there are still anything should improved here

amanda-anderson commented:
@amirabayoumi I'll do another code review today! :raised_hands:

amirabayoumi committed:
resolve the conflict

amanda-anderson reviewed (approved):
Looking great! :rocket: I think it's ready to go! :clap: 

Just make sure to run the E2E tests once if you haven't already before merging to master :relaxed:

IvaKop reviewed (approved):
Great work, Amira! 🚀

amirabayoumi committed:
resolve the conflict

Commit:
Translations updated by CI (extract-intl)

amirabayoumi committed:
Merge branch 'master' into TAN-5788-Complex-graphics-for-Accessibility

amirabayoumi committed:
add missing curly braces

amirabayoumi committed:
Merge branch 'TAN-5788-Complex-graphics-for-Accessibility' of https://github.com/CitizenLabDotCo/citizenlab into TAN-5788-Complex-graphics-for-Accessibility

Commit:
Translations updated by CI (extract-intl)
